support-services/src/main/scala/com/gu/support/promotions/PromotionService.scala

```diff
@@ -26,10 +26,12 @@ class PromotionService(config: PromotionsConfig, maybeCollection: Option[Promoti
     }
 
   def findPromotions(promoCodes: List[PromoCode]): List[PromotionWithCode] =
-    allWith6For6
-      .foldLeft(List.empty[PromotionWithCode]) { (acc, promotion) =>
-        val maybeCode = promoCodes.intersect(promotion.promoCodes.toList).headOption
-        maybeCode.map(code => acc :+ PromotionWithCode(code, promotion)).getOrElse(acc)
+    promoCodes
```

Comment:
re duplicates, I think that was tightened up after the infamous ONEFORONE incident?  There was a card for it - whether it was done or not I'm not sure.

Reply:
I don't know about that incident! But the model still allows many codes per promo

Reply:
if there are duplicates even your current solution would break compared with the old one.  You would need to use `filter` instead of `find` on line 31 to get the old behaviour

Reply:
however due to the headOption, the old one would only get one promo code from each promotion.  So if a promotion had AAA and BBB, and the promoCodes parameter had both AAA and BBB, the old implementation would only add one of them to the list, but the new one would add AAA and BBB.  Not sure if that's an issue.

Reply:
I wrote this above comment before your later comments arrived!

Reply:
Re the full list scan, the promotions are held in memory so I wouldn't have any concerns about performance

Reply:
>The old logic used headOption, so having a lookup by promoCode wouldn't be much different - either way only 1 promo per promoCode is supported
jsut to be clear, I think the old solution would handle a duplicate promo code in multiple promotions, and pick them all up.  The new one would just pick the first promotion.
secondly, the old solution would only pick up the first matching promo code within each promotion, whereas the new solution would pick the same promotion up for each promo code within that promotion that was passed in the input.

Reply:
I see what you mean. If a promoCode appears across multiple promotions then we currently take them all. Perhaps we don't want to change this behavour, I'll have a think...

Reply:
Ok, I've pushed a commit which preserved the existing behaviour - it supports more than one promotion per promoCode. It's about as efficient as the original solution - this time it traverses `allWith6For6` once to build a `Map`. Really this could be memoized in the `promotionCollection`, but that's an unnecessary optimisation.